source/RoslynAnalyzers/MessageContractAnalyzers.cs

```diff
@@ -0,0 +1,222 @@
+                result &= MessageTypes_MustInstantiateCollections(context, propDec, required, isCollectionType);
+                result &= PropertiesOnMessageTypes_MustHaveAtLeastOneValidationAttribute(context, propDec, required);
+                result &= SpaceIdPropertiesOnMessageTypes_MustBeOfTypeSpaceId(context, propDec);
+                result &= IdPropertiesOnMessageTypes_MustBeACaseInsensitiveStringTinyType(context, propDec);
+            }
+
+            return result;
+        }
+        
+        // ----- helpers --------------- 
+        
+        enum RequiredState
+        {
+            Unspecified,
+            Optional,
+            Required,
+        }
+
+        static RequiredState GetRequiredState(PropertyDeclarationSyntax propDec)
+        {
+            var attrNames = propDec.AttributeLists.SelectMany(al => al.Attributes.Select(a => (a.Name as IdentifierNameSyntax)?.Identifier.Text));
+
+            foreach (var attrName in attrNames)
+            {
+                // it has an attribute called Required. Not necessarily the one from System.ComponentModel.DataAnnotations but good enough.
+                // Note: we could resolve the full type which is more expensive, it's unclear as to whether that makes a difference or not
+                if (attrName == "Required") return RequiredState.Required;
+                // it has an attribute called Optional. Not necessarily the one from Octopus.MessageContracts.Attribute but good enough.
+                if (attrName == "Optional") return RequiredState.Optional;
+            }
+
+            return RequiredState.Unspecified;
+        }
+
+        static bool IsCollectionType(SyntaxNodeAnalysisContext context, PropertyDeclarationSyntax propDec)
+        {
+            var propType = propDec.Type is NullableTypeSyntax n ? n.ElementType : propDec.Type;
+
+            var typeInfo = context.SemanticModel.GetTypeInfo(propType);
+            if (typeInfo.Type == null) return false;
+
+            return typeInfo.Type.IsAssignableTo(cachedTypes.IEnumerable) && !SymbolEqualityComparer.Default.Equals(typeInfo.Type, cachedTypes.String);
+        }
+       
+        // "Request" might be part of the name of the request DTO, so we only want to replace the last occurrence of the word "Request"
+        static string ReplaceLast(string str, string oldValue, string newValue)
+        {
+            var pos = str.LastIndexOf(oldValue, StringComparison.Ordinal);
+            return pos != -1 ? str.Remove(pos, oldValue.Length).Insert(pos, newValue) : str;
+        }
+        
+        // Roslyn deals in MSIL types rather than C# types, so we get Int32 rather than int.
+        // this converts back to language-specific aliases.
+        static string CSharpNameForType(ITypeSymbol? symbol)
+        {
+            var symName = symbol?.Name;
+            // there is probably some builtin method to do this reverse lookup; replace this switch if you find out how
```

Comment:
[Apparently not ](https://stackoverflow.com/a/1362899/10784)

Reply:
Yuck. Thanks for the link; I'll bring in the full table from there